examples/deal.II/bps.h

```diff
@@ -0,0 +1,878 @@
+    build_ctx_data.space_dim = dim;
+
+    CeedQFunctionContextCreate(ceed, &build_ctx);
+    CeedQFunctionContextSetData(
+      build_ctx, CEED_MEM_HOST, CEED_USE_POINTER, sizeof(build_ctx_data), &build_ctx_data);
+
+    // 5) create q operation
+    if (bp == BPType::BP1)
+      CeedQFunctionCreateInterior(ceed, 1, f_apply_mass, f_apply_mass_loc, &qf_apply);
+    else if (bp == BPType::BP2)
+      CeedQFunctionCreateInterior(ceed, 1, f_apply_mass_vec, f_apply_mass_vec_loc, &qf_apply);
+    else if (bp == BPType::BP3 || bp == BPType::BP5)
+      CeedQFunctionCreateInterior(ceed, 1, f_apply_poisson, f_apply_poisson_loc, &qf_apply);
+    else if (bp == BPType::BP4 || bp == BPType::BP6)
+      CeedQFunctionCreateInterior(ceed, 1, f_apply_poisson_vec, f_apply_poisson_vec_loc, &qf_apply);
+    else
+      AssertThrow(false, ExcInternalError());
+
+    if (bp <= BPType::BP2)
+      CeedQFunctionAddInput(qf_apply, "u", n_components, CEED_EVAL_INTERP);
+    else
+      CeedQFunctionAddInput(qf_apply, "u", dim * n_components, CEED_EVAL_GRAD);
+
+    CeedQFunctionAddInput(qf_apply, "qdata", n_components_metric, CEED_EVAL_NONE);
+
+    if (bp <= BPType::BP2)
+      CeedQFunctionAddOutput(qf_apply, "v", n_components, CEED_EVAL_INTERP);
+    else
+      CeedQFunctionAddOutput(qf_apply, "v", dim * n_components, CEED_EVAL_GRAD);
+
+    CeedQFunctionSetContext(qf_apply, build_ctx);
+
+    // 6) put everything together
+    CeedOperatorCreate(ceed, qf_apply, CEED_QFUNCTION_NONE, CEED_QFUNCTION_NONE, &op_apply);
+
+    CeedOperatorSetField(op_apply, "u", sol_restriction, sol_basis, CEED_VECTOR_ACTIVE);
+    CeedOperatorSetField(op_apply, "qdata", q_data_restriction, CEED_BASIS_COLLOCATED, q_data);
+    CeedOperatorSetField(op_apply, "v", sol_restriction, sol_basis, CEED_VECTOR_ACTIVE);
+  }
+
+  /**
+   * Perform matrix-vector product.
+   */
+  void
+  vmult(VectorType &dst, const VectorType &src) const override
+  {
+    // communicate: update ghost values
+    src.update_ghost_values();
+
+    if (dof_handler.get_fe().n_components() == 1)
+      {
+        // create libCEED view on deal.II vectors
+        VectorTypeCeed src_ceed(ceed, src);
+        VectorTypeCeed dst_ceed(ceed, dst);
+
+        // apply operator
+        CeedOperatorApply(op_apply, src_ceed(), dst_ceed(), CEED_REQUEST_IMMEDIATE);
+      }
+    else // TODO: needed for multiple components
+      {
+        // allocate space for block vectors
+        src_tmp.reinit(this->extended_local_size(), true);
+        dst_tmp.reinit(this->extended_local_size(), true);
+
+        // create libCEED view on deal.II vectors
+        VectorTypeCeed src_ceed(ceed, src_tmp);
+        VectorTypeCeed dst_ceed(ceed, dst_tmp);
+
+        copy_to_block_vector(src_tmp, src); // copy to block vector
+
+        // apply operator
+        CeedOperatorApply(op_apply, src_ceed(), dst_ceed(), CEED_REQUEST_IMMEDIATE);
+
+        copy_from_block_vector(dst, dst_tmp); // copy from block vector
+      }
```

Comment:
Ah, I think the issue is here. This seems to fix it, though I'm not sure if there is a more idiomatic approach

```diff
$ git diff
diff --git a/examples/deal.II/bps.h b/examples/deal.II/bps.h
index cfde7f7f..7a76f985 100644
--- a/examples/deal.II/bps.h
+++ b/examples/deal.II/bps.h
@@ -325,14 +325,17 @@ public:
         src_tmp.reinit(this->extended_local_size(), true);
         dst_tmp.reinit(this->extended_local_size(), true);
 
-        // create libCEED view on deal.II vectors
-        VectorTypeCeed src_ceed(ceed, src_tmp);
-        VectorTypeCeed dst_ceed(ceed, dst_tmp);
+        // copy to block vector
+        copy_to_block_vector(src_tmp, src);
 
-        copy_to_block_vector(src_tmp, src); // copy to block vector
+        {
+          // create libCEED view on deal.II vectors
+          VectorTypeCeed src_ceed(ceed, src_tmp);
+          VectorTypeCeed dst_ceed(ceed, dst_tmp);
 
-        // apply operator
-        CeedOperatorApply(op_apply, src_ceed(), dst_ceed(), CEED_REQUEST_IMMEDIATE);
+          // apply operator
+          CeedOperatorApply(op_apply, src_ceed(), dst_ceed(), CEED_REQUEST_IMMEDIATE);
+        }
 
         copy_from_block_vector(dst, dst_tmp); // copy from block vector
       }
@@ -416,6 +419,8 @@ private:
      */
     ~VectorTypeCeed()
     {
+      CeedScalar *ptr;
+      CeedVectorTakeArray(vec_ceed, CEED_MEM_HOST, &ptr);
       CeedVectorDestroy(&vec_ceed);
     }
```